core/os/android/adb/commands.go

```diff
@@ -256,6 +261,51 @@ func (b *binding) SupportsPerfetto(ctx context.Context) bool {
 	return os.GetAPIVersion() >= 28
 }
 
+func (b *binding) QueryPerfettoServiceState(ctx context.Context) (*device.PerfettoCapability, error) {
+	if b.Instance().GetConfiguration().GetOS().GetAPIVersion() <= 29 {
+		return &device.PerfettoCapability{}, log.Errf(ctx, nil, "Querying perfetto capability requires Android API > 29")
+	}
+	res, err := b.Shell("perfetto", "--query-raw", "|", "base64").Call(ctx)
+	if err != nil {
+		return &device.PerfettoCapability{}, log.Errf(ctx, err, "adb shell perfetto returned error: %s", res)
+	}
+	decoded, _ := base64.StdEncoding.DecodeString(res)
+	tracingServiceState := &perfetto_pb.TracingServiceState{}
+	if err = proto.Unmarshal(decoded, tracingServiceState); err != nil {
+		return &device.PerfettoCapability{}, log.Errf(ctx, err, "Unmarshal returned error")
+	}
+	perfettoCapability := &device.PerfettoCapability{
+		GpuProfiling: &device.GPUProfiling{
+			HasRenderStage:       false,
+			GpuCounterDescriptor: nil,
+		},
+	}
+	dataSources := tracingServiceState.GetDataSources()
+	for _, dataSource := range dataSources {
+		dataSourceDescriptor := dataSource.GetDsDescriptor()
+		if dataSourceDescriptor.GetName() == GPURENDERINGSTAGESDATASOURCEDESCRIPTORNAME {
+			perfettoCapability.GpuProfiling.HasRenderStage = true
+			continue
+		}
+		gpuCounterDescriptor := dataSourceDescriptor.GetGpuCounterDescriptor()
+		specs := gpuCounterDescriptor.GetSpecs()
+		if len(specs) == 0 || perfettoCapability.GpuProfiling.GpuCounterDescriptor != nil {
+			continue
+		}
+		data, err := proto.Marshal(gpuCounterDescriptor)
```

Comment:
Please add a comment here that we're going via the binary form, because we expect them to be the same.